components/search/src/filter.rs

```diff
@@ -217,4 +299,458 @@ mod tests {
+                    environment: JSONVariantEnvironment {
+                        all_regions_and_locales: false,
+                        excluded_locales: vec![],
+                        excluded_regions: vec![],
+                        locales: vec!["en-GB".to_string()],
+                        regions: vec![],
+                        distributions: vec![],
+                        excluded_distributions: vec![],
+                        applications: vec![],
+                    },
+                }],
+            }),
+            &SearchUserEnvironment {
+                locale: "fi".into(),
+                region: "FR".into(),
+                update_channel: SearchUpdateChannel::Default,
+                distribution_id: String::new(),
+                experiment: String::new(),
+                app_name: SearchApplicationName::Firefox,
+                version: String::new(),
+            },
+        );
+
+        assert_eq!(
+            default_engine_id.unwrap(),
+            "engine2",
+            "Should have returned the global default engine when no specific defaults environments match"
+        );
+        assert_eq!(
+            default_engine_private_id.unwrap(),
+            "engine3",
+            "Should have returned the global default engine for private mode when no specific defaults environments match"
+        );
+
+        let (default_engine_id, default_engine_private_id) = determine_default_engines(
+            &ENGINES_LIST,
+            Some(JSONDefaultEnginesRecord {
+                global_default: "engine2".to_string(),
+                global_default_private: "engine3".to_string(),
+                specific_defaults: vec![JSONSpecificDefaultRecord {
+                    default: String::new(),
+                    default_private: "engine1".to_string(),
+                    environment: JSONVariantEnvironment {
+                        all_regions_and_locales: false,
+                        excluded_locales: vec![],
+                        excluded_regions: vec![],
+                        locales: vec!["fi".to_string()],
+                        regions: vec![],
+                        distributions: vec![],
+                        excluded_distributions: vec![],
+                        applications: vec![],
+                    },
+                }],
+            }),
+            &SearchUserEnvironment {
+                locale: "fi".into(),
+                region: "FR".into(),
+                update_channel: SearchUpdateChannel::Default,
+                distribution_id: String::new(),
+                experiment: String::new(),
+                app_name: SearchApplicationName::Firefox,
+                version: String::new(),
+            },
+        );
+
+        assert_eq!(
+            default_engine_id.unwrap(),
+            "engine2",
+            "Should have returned the global default engine when specific environments match, but do not override the global"
```

Comment:
I understand we're testing the default_engine_id is "engine2" but how does this test make sure global default `engine2` is not overridden? 

Same comment for the test below.

Reply:
Actually, I'm still not understanding this assertion statement because I don't understand why or how the global engine would ever become overridden?

Reply:
We're trying to test that if the specific environment is overriding the private default, then it is not overriding the global default as well.

I've updated the wording for that.